test_utils/test_utils/paho_fuzz_hook.py

```diff
@@ -0,0 +1,249 @@
+        """
+        nonlocal failed
+        if not failed and random.random() < failure_probability:
+            log_func("-----------SOCKET FAILURE. All incoming packets will be dropped")
+            failed = True
+
+        buf = old_sock_recv(buffsize)
+
+        if failed:
+            log_func("-----------DROPPING {} bytes".format(len(buf)))
+            raise BlockingIOError
+        else:
+            log_func("---------- RECEIVED {} bytes".format(len(buf)))
+            return buf
+
+    @functools.wraps(old_sock_close)
+    def new_sock_close():
+        """
+        Inside `sock_close`, we reset the `failed` variable to `False` to simulate the connection
+        being "fixed" after the socket is re-opened
+        """
+        nonlocal failed
+        if failed:
+            log_func("-----------RESTORING SOCKET behavior")
+            failed = False
+        return old_sock_close()
+
+    paho._sock_recv = new_sock_recv
+    paho._sock_close = new_sock_close
+
+
+def add_hook_drop_individual_incoming(device_client, failure_probability, log_func=print):
+    """
+    Add a hook to randomly drop individual incoming messages until reconnect based on some
+    failure probability. This is used to simulate a "unreliable network" scenario where individual
+    outgoing packets get dropped, but other packets continue to flow.  Since we don't know what
+    an "individual message" is when we're receiving, we just flush the incoming byte queue and
+    assume that is good enough.
+    """
+    paho = logging_hook.get_paho_from_device_client(device_client)
+
+    old_sock_recv = paho._sock_recv
+
+    @functools.wraps(old_sock_recv)
+    def new_sock_recv(buffsize):
+        """
+        Inside `sock_recv` we randomly flush the incoming packet queue to simulate dropping of
+        a single packet, (or a small number of incoming packets depending on how quickly the
+        incoming queue is handled).
+        """
+        if random.random() < failure_probability:
+            buf = old_sock_recv(2048)
+            log_func("---------- DROPPED {} bytes".format(len(buf)))
+            raise BlockingIOError
+        else:
+            buf = old_sock_recv(buffsize)
+            log_func("---------- RECEIVED {} bytes".format(len(buf)))
+            return buf
+
+    paho._sock_recv = new_sock_recv
+
+
+def add_hook_raise_send_exception(device_client, failure_probability, log_func=print):
+    """
+    Add a hook to randomly raise an exception when sending based on some failure probability. This
+    is used to simulate an exception inside Paho when sending.
+    """
+    paho = logging_hook.get_paho_from_device_client(device_client)
+
+    old_sock_send = paho._sock_send
+
+    @functools.wraps(old_sock_send)
+    def new_sock_send(buf):
+        if random.random() < failure_probability:
+            log_func("---------- RAISING EXCEPTION")
+            raise Exception("Forced Send Failure")
```

Comment:
Once again, here and everywhere I'm concerned about the use of Exception as opposed to a custom defined exception that inherits from it